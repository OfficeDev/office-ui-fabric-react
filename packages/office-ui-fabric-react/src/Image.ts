export * from './packages/react-primitives/Image';
