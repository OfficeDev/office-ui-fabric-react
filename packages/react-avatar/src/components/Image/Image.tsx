export { ImageBase as Image } from './Image.base';
