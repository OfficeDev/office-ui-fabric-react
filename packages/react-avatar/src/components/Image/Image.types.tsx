import { ComponentProps } from '../utils/commonTypes';
import { ComposeOptions, ComposeStandardStatics } from '../utils/compose';

// tslint:disable-next-line:no-any
export interface ImageProps extends ComponentProps, React.ImgHTMLAttributes<HTMLImageElement> {}

export interface ImageOptions extends ComposeOptions<ImageProps, {}, {}, ComposeStandardStatics> {}
