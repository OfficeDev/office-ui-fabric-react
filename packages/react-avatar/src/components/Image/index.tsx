export * from './Image.base';
export * from './Image.types';
export * from './useImage';
export * from './Image';
