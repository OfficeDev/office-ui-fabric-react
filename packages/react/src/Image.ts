import { version } from 'process';

export * from './components/Image/index';
