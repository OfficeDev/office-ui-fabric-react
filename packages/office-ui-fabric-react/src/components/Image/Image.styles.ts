import { ImageFit, IImageStyleProps, IImageStyles, ImageLoadState } from './Image.types';
import { IStyle } from '../../Styling';
import { AnimationClassNames } from '../../Styling';

const OneHundredPercent = '100%';
const Auto = 'auto';

export const getStyles = (props: IImageStyleProps): IImageStyles => {
  const {
    className,
    imageFit,
    loadState,
    isPortrait,
    maximizeFrame,
    width,
    height,
    shouldFadeIn
  } = props;
  const isLoaded = loadState === ImageLoadState.loaded;
  const isError = loadState === ImageLoadState.error;
  const isBestFit = imageFit === undefined;
  const scaleWidth = isBestFit && width !== undefined && height === undefined;
  const scaleHeight = isBestFit && width === undefined && height !== undefined;
  const scaleWidthHeight = isBestFit && width !== undefined && height !== undefined;

  return {

    root: [
      'ms-Image',
      isPortrait && 'ms-Image-image--portrait',
      !isPortrait && 'ms-Image-image--landscape',
      {
        overflow: 'hidden'
      },
      maximizeFrame && {
        height: OneHundredPercent,
        width: OneHundredPercent
      },
      className
    ],

    image: [
      'ms-Image-image',
      imageFit === ImageFit.none && 'ms-Image-image--none',
      imageFit === ImageFit.center && 'ms-Image-image--center',
      imageFit === ImageFit.contain && 'ms-Image-image--contain',
      imageFit === ImageFit.cover && 'ms-Image-image--cover',
      !isLoaded && 'is-notLoaded',
      isLoaded && 'is-loaded ',
      shouldFadeIn && 'is-fadeIn',
      loadState === ImageLoadState.error && 'is-error',
      scaleWidth && 'ms-Image-image--scaleWidth',
      scaleHeight && 'ms-Image-image--scaleHeight',
      scaleWidthHeight && 'ms-Image-image--scaleWidthHeight',

      {
        display: 'block',
        opacity: 0
      },
      isLoaded && {
        opacity: 1
      },
      isLoaded && shouldFadeIn && AnimationClassNames.fadeIn400,
      (imageFit === ImageFit.center ||
        imageFit === ImageFit.contain ||
        imageFit === ImageFit.cover
      ) && {
        position: 'relative',
        left: '50%',
        top: '50%',
        transform: 'translate(-50%, -50%)'
      },
      imageFit === ImageFit.contain && [
        !isPortrait && {
          width: OneHundredPercent,
          height: Auto
        },
        isPortrait && {
          width: Auto,
          height: OneHundredPercent
        }
      ],
      imageFit === ImageFit.cover && [
        !isPortrait && {
          height: OneHundredPercent,
          width: Auto
        },
        isPortrait && {
          width: OneHundredPercent,
          height: Auto
        }
      ],
      scaleWidthHeight && {
        width: OneHundredPercent,
        height: OneHundredPercent
      },
      scaleWidth && {
        width: OneHundredPercent,
        height: Auto
      },
      scaleHeight && {
        width: Auto,
        height: OneHundredPercent
      }
    ]

  };
};
