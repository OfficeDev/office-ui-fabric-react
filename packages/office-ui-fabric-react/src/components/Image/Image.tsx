import { ImageBase } from './Image.base';
import { IImageProps } from './Image.types';
import { getStyles } from './Image.styles';
import { styled } from '../../Styling';

export const Image = styled(
  ImageBase,
  getStyles
);
