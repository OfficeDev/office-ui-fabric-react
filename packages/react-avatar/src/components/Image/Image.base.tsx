import * as React from 'react';
import { ImageProps, ImageOptions } from './Image.types';
import { compose, ComposeStandardStatics } from '../utils/compose';
import { useImage } from './useImage';

export const ImageBase = compose<ImageProps, {}, {}, ComposeStandardStatics>(
  (props: ImageProps & ImageOptions, ref: React.RefObject<HTMLElement>, options: ImageOptions) => {
    const { slots, slotProps } = useImage(props, options);

    return <slots.root ref={ref} {...slotProps.root} />;
  },
  {
    defaultProps: {
      as: 'img',
    },
    statics: {
      handledProp: 'src',
      displayName: 'ImageBase',
    },
  },
);
