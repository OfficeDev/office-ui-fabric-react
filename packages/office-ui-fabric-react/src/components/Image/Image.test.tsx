/* tslint:disable:no-unused-variable */
import * as React from 'react';
/* tslint:enable:no-unused-variable */

import * as ReactDOM from 'react-dom';
import * as ReactTestUtils from 'react-addons-test-utils';
import { shallow, mount } from 'enzyme';

import { Image } from './Image';
import { ImageFit, ImageLoadState, ImageCoverStyle } from './Image.Props';

/* tslint:disable:no-unused-variable */
const testImage1x1 = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAC0lEQVQImWP4DwQACfsD/eNV8pwAAAAASUVORK5CYII=';
const testImage1x2 = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAACCAYAAACZgbYnAAAAEklEQVQImWP4////fyYGBgYGAB32A/+PRyXoAAAAAElFTkSuQmCC';
const testImage2x1 = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAIAAAABCAYAAAD0In+KAAAAEUlEQVQImWP8////fwYGBgYAGfgD/hEzDhoAAAAASUVORK5CYII=';
const brokenImage = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgFcSJAAAAC0lEQVQImWP4DwQACfsD/eNV8pwAAAAASUVORK5CYII=';
/* tslint:enable:no-unused-variable */

describe('Image', () => {

  it('renders an image', (done) => {
    let component = ReactTestUtils.renderIntoDocument(
      <Image
        src={ testImage1x1 }
        onLoad={ () => done() }
      />
    );

    let image = ReactTestUtils.findRenderedDOMComponentWithTag(component as any, 'img');
    ReactTestUtils.Simulate.load(image);

  });

  it.only('can cover a portrait (tall) frame with a square image', (done) => {
    // let root = document.createElement('div');
    // let onLoadingStateChange = (loadState: ImageLoadState) => {
    //   if (loadState === ImageLoadState.loaded) {
    //     let image = document.querySelector('.ms-Image.is-portraitFrame .ms-Image-image') as Element;
    //     try {
    //       expect(image.className).to.contain('ms-Image-image--landscape');
    //     } catch (e) { done(e); }
    //     done();
    //   }
    // };
    // document.body.appendChild(root);
    // ReactDOM.render<HTMLDivElement>(
    //   <Image
    //     src={ testImage1x1 }
    //     width={ 1 }
    //     height={ 3 }
    //     imageFit={ ImageFit.cover }
    //     className='is-portraitFrame'
    //     onLoadingStateChange={ onLoadingStateChange }
    //   />, root
    // );

    let component = mount(
      <div>
        <Image
          src={ testImage1x1 }
          width={ 1 }
          height={ 3 }
          imageFit={ ImageFit.cover }
          className='is-portraitFrame'
          onLoadingStateChange={ () => {
            // console.log('ONLOADCHANGE', component.find('img').getNode());
            expect(component.find('.ms-Image-image--landscape')).toHaveLength(1);
            done();
          } }
        />
      </div>
    );
   // console.log('AFTER MOUNT', component.find('img').getNode());

    component.find('img').simulate('load');
  });

  it('can cover a landscape (wide) frame with a square image', () => {
    let component = mount(
      <div>
        <Image
          src={ testImage1x1 }
          width={ 3 }
          height={ 1 }
          imageFit={ ImageFit.cover }
          className='is-landscapeFrame'

        />
      </div>
    );

    expect(component.find('.ms-Image-image--portrait')).toHaveLength(1);
  });

  it('can cover a landscape (wide) parent element with a square image', () => {
    // let root = document.createElement('div');
    // let onLoadingStateChange = (loadState: ImageLoadState) => {
    //   if (loadState === ImageLoadState.loaded) {
    //     let image = document.querySelector('.ms-Image.is-frameMaximizedPortrait .ms-Image-image') as Element;
    //     try {
    //       expect(image.className).to.contain('ms-Image-image--portrait');
    //     } catch (e) { done(e); }
    //     done();
    //   }
    // };
    // document.body.appendChild(root);
    // ReactDOM.render<HTMLDivElement>(
    //   <div style={ { width: '20px', height: '10px' } }>
    //     <Image
    //       className='is-frameMaximizedPortrait'
    //       imageFit={ ImageFit.cover }
    //       maximizeFrame
    //       src={ testImage1x1 }
    //       onLoadingStateChange={ onLoadingStateChange }
    //     />
    //   </div>, root
    // );

    let component = mount(
      <div style={ { width: '20px', height: '10px' } }>
        <Image
          className='is-frameMaximizedPortrait'
          imageFit={ ImageFit.cover }
          maximizeFrame
          src={ testImage1x1 }
        />
      </div>);

    expect(component.find('.ms-Image-image--landscape')).toHaveLength(1);
  });

  it('can cover a portrait (tall) parent element with a square image', (done) => {
    // let root = document.createElement('div');
    // let onLoadingStateChange = (loadState: ImageLoadState) => {
    //   if (loadState === ImageLoadState.loaded) {
    //     let image = document.querySelector('.ms-Image.is-frameMaximizedLandscape .ms-Image-image') as Element;
    //     try {
    //       expect(image.className).to.contain('ms-Image-image--landscape');
    //     } catch (e) { done(e); }
    //     done();
    //   }
    // };
    // document.body.appendChild(root);
    // ReactDOM.render<HTMLDivElement>(
    //   <div style={ { width: '10px', height: '20px' } }>
    //     <Image
    //       src={ testImage1x1 }
    //       imageFit={ ImageFit.cover }
    //       className='is-frameMaximizedLandscape'
    //       onLoadingStateChange={ onLoadingStateChange }
    //       maximizeFrame
    //     />
    //   </div>, root
    // );

    let component = mount(
      <div style={ { width: '10px', height: '20px' } }>
        <Image
          src={ testImage1x1 }
          imageFit={ ImageFit.cover }
          className='is-frameMaximizedLandscape'
          maximizeFrame
        />
      </div>
    );

    expect(component.find('.ms-Image-image--landscape')).toHaveLength(1);
  });

  it('allows onError events to be attached', (done) => {
    let component = ReactTestUtils.renderIntoDocument(
      <Image
        src={ brokenImage }
        onError={ () => done() }
      />
    );

    let img = ReactTestUtils.findRenderedDOMComponentWithTag(component as any, 'img');
    ReactTestUtils.Simulate.error(img);
  });
});
