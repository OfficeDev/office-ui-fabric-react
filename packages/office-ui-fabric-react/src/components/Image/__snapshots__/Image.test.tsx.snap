// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders Image correctly 1`] = `
<div
  className=
      ms-Image
      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        overflow: hidden;
      }
  style={
    Object {
      "height": undefined,
      "width": undefined,
    }
  }
>
  <img
    className=
        ms-Image-image
        ms-Image-image--portrait
        is-notLoaded
        is-fadeIn
        {
          display: block;
          opacity: 0;
        }
    onError={[Function]}
    onLoad={[Function]}
    src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAC0lEQVQImWP4DwQACfsD/eNV8pwAAAAASUVORK5CYII="
  />
</div>
`;
